### ISSUE TITLE
Does DatabaseConnection::exec() need to be &mut self ?

### ISSUE THREAD

**Issue by ericsink** [dckc/rust-sqlite3#40]
Life gets much better if I change this to:

```
pub fn exec(&self, sql: &str) -> SqliteResult<()>
```

The consistency-based argument here is that exec() and prepare() should be the same.

The not-very-compelling argument is that exec(&mut self) means that I can only call exec() from one reference and that's a huge pain.

But I'll admit that whenever I think about the boundary between "safety assurances provided by SQLite" and "additional safety assurances provided by Rust", I get myself tied up in knots.

**Comment by dckc** [comment]
I wrote it that way because I didn't have any assurance that `sqlite3_exec` didn't mutate the underlying `sqlite3`. @PeterReid, thoughts?

I suppose that applies to prepare() as well, though I thought of prepare() as creating a new sqlite3_stmt without modifying the `sqlite3`.

Care to elaborate on "huge pain", @ericsink?

**Comment by ericsink** [comment]
I'm just not sure what "mutating the underlying `sqlite3`" means.

Is it different from "mutating one or more of the database files which are attached to that `sqlite3`"?  Thinking about trying to express that in Rust's type system seems hopeless.

**Comment by ericsink** [comment]
I'll get back to you on "huge pain".  I'm still trying to get my story straight.

**Comment by dckc** [comment]
tldr: hellifino

I tried to consult the gods in the #rust IRC channel, but they're not handy just now.

thinking out loud...

by "mutating the underlying sqlite3" I mean: updating that data structure or any memory it refers to, directly or indirectly. So in the case of an in-memory database, yes, database changes count.

You saw the suggestion in #37 to use thread-safe connections. Something tells me that would take me into `Arc<RefCell<X>>` territory, which I have been avoiding. Maybe there should be 2 core APIs: one Sync and one not Sync? Ugh. I dunno.

**Comment by ericsink** [comment]
Yeah, I got no clue what the right answer is here.  I'm going to keep going forward, trying stuff, and see what gets me into trouble.

For now, a little more thinking out loud:

The issue with mut in Rust is that it doesn't just block other mutators (writers).  It also blocks all readers.  That's great, but it doesn't match SQLite's semantics.

Actually, for SQLite's "delete" journal mode, this matches fine.  But for WAL mode, it does not.  In WAL mode, there can be only one writer, but writers do not block readers, and readers do not block writers.  AFAIK, Rust doesn't have a pointer which works this way.  But I should probably re-read this yet again:

http://manishearth.github.io/blog/2015/05/27/wrapper-types-in-rust-choosing-your-guarantees/

Anyway, I'm just stewing about how much database locking semantic stuff can or should be mapped onto the Rust type system.

**Comment by dckc** [comment]
This is now Somebody Else's Problem :)
See #46
